ussomacaroon/client.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Pacakge ussomacaroon provides a client that can authenticate with an
+// identity server by discharging macaroons on an Ubuntu SSO server.
+package ussomacaroon
+
+import (
+	"net/url"
+
+	"github.com/juju/httprequest"
+	errgo "gopkg.in/errgo.v1"
+	"gopkg.in/macaroon-bakery.v2-unstable/bakery"
+	"gopkg.in/macaroon-bakery.v2-unstable/httpbakery"
+	"gopkg.in/macaroon.v2-unstable"
+)
+
+// Macaroon returns a macaroon from the identity provider at the given
+// URL, which will have a third-party caveat
+// addressed to an Ubuntu SSO service.
+func Macaroon(url string) (*macaroon.Macaroon, error) {
+	client := &httprequest.Client{}
+	var resp MacaroonResponse
+	if err := client.Get(url, &resp); err != nil {
+		return nil, errgo.Notef(err, "cannot get macaroon")
+	}
+	return resp.Macaroon, nil
+}
+
+// Visitor is an httpbakery.Visitor that will login using a macaroon
+// discharged by an Ubuntu SSO service.
+type Visitor struct {
+	f func(string) (macaroon.Slice, error)
+}
+
+// NewVisitor creates a Vistor which uses a macaroon previously collected
```

Comment:
I found this a little confusing at first.

How about something like:

    NewVisitor returns a visitor that calls f to obtain a macaroon
    which is used to complete a discharge request.
    The argument to f is a URL suitable for passing to the Macaroon
    function.

?